cpp/src/arrow/python/common.h

```diff
@@ -148,41 +148,59 @@ struct ARROW_EXPORT PyBytesView {
   // View the given Python object as binary-like, i.e. bytes
   Status FromBinary(PyObject* obj) { return FromBinary(obj, "a bytes object"); }
 
-  // View the given Python object as string-like, i.e. str or (utf8) bytes
-  Status FromString(PyObject* obj, bool check_valid = false) {
-    if (PyUnicode_Check(obj)) {
+  Status FromString(PyObject* obj) {
+    bool ignored = false;
+    return FromString(obj, false, &ignored);
+  }
+
+  Status FromString(PyObject* obj, bool* is_utf8) {
+    return FromString(obj, true, is_utf8);
+  }
+
+  Status FromUnicode(PyObject* obj) {
 #if PY_MAJOR_VERSION >= 3
-      Py_ssize_t size;
-      // The utf-8 representation is cached on the unicode object
-      const char* data = PyUnicode_AsUTF8AndSize(obj, &size);
-      RETURN_IF_PYERROR();
-      this->bytes = data;
-      this->size = size;
-      this->ref.reset();
-      return Status::OK();
+    Py_ssize_t size;
+    // The utf-8 representation is cached on the unicode object
+    const char* data = PyUnicode_AsUTF8AndSize(obj, &size);
+    RETURN_IF_PYERROR();
+    this->bytes = data;
+    this->size = size;
+    this->ref.reset();
 #else
-      PyObject* converted = PyUnicode_AsUTF8String(obj);
-      RETURN_IF_PYERROR();
-      this->bytes = PyBytes_AS_STRING(converted);
-      this->size = PyBytes_GET_SIZE(converted);
-      this->ref.reset(converted);
-      return Status::OK();
+    PyObject* converted = PyUnicode_AsUTF8String(obj);
+    RETURN_IF_PYERROR();
+    this->bytes = PyBytes_AS_STRING(converted);
+    this->size = PyBytes_GET_SIZE(converted);
+    this->ref.reset(converted);
 #endif
+    return Status::OK();
+  }
+
+ protected:
+  PyBytesView(const char* b, Py_ssize_t s, PyObject* obj = NULLPTR)
+      : bytes(b), size(s), ref(obj) {}
+
+  // View the given Python object as string-like, i.e. str or (utf8) bytes
+  Status FromString(PyObject* obj, bool check_utf8, bool* is_utf8) {
+    if (PyUnicode_Check(obj)) {
+      *is_utf8 = true;
```

Comment:
Is `is_utf8` supposed to be filled in when `check_utf8` is false?

Reply:
Fixed